File: src/chipschallenge/blockreactions/YellowKeyTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Inventory.Key;

/**
 *
 * @author patrik
 */
public class YellowKeyTo extends BlockReaction {

    private YellowKeyTo() {
    }
    private static YellowKeyTo mInstance = null;

    public static synchronized YellowKeyTo getInstance() {
        if (mInstance == null) {
            mInstance = new YellowKeyTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            takeKey(Key.YELLOW);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/YellowKeyTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;
    using Type = Block.Type;

    public class YellowKeyTo : NoSlipReaction
    {
        private static YellowKeyTo instance;

        private YellowKeyTo()
        {
        }

        public static YellowKeyTo Instance
        {
            get
            {
                lock (typeof(YellowKeyTo))
                {
                    if (instance == null)
                    {
                        instance = new YellowKeyTo();
                    }
                    return instance;
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeKey(Key.YELLOW);
                standing.Replace(CreateBlock(Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}